### PR TITLE
Add animation and transition support for pseudo-elements

### LINKED ISSUE #10316: Support pseudo elements for transition and animations
``` html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <title></title>

    <style>
        /* Sweep To Right */      
        .hvr-sweep-to-right {
            display: inline-block;
            vertical-align: middle;
            position: relative;
            -webkit-transition-property: color;
            transition-property: color;
            -webkit-transition-duration: 0.3s;
            transition-duration: 0.3s;
        }
        .hvr-sweep-to-right:before {
            content: "";
            position: absolute;
            z-index: -1;
            top: 0;
            left: 0;
            right: 0;
            bottom: 0;
            background: #2098d1;
            -webkit-transform: scaleX(0);
            transform: scaleX(0);
            -webkit-transform-origin: 0 50%;
            transform-origin: 0 50%;
            -webkit-transition-property: transform;
            transition-property: transform;
            -webkit-transition-duration: 0.3s;
            transition-duration: 0.3s;
            -webkit-transition-timing-function: ease-out;
            transition-timing-function: ease-out;
        }
        .hvr-sweep-to-right:hover,
        .hvr-sweep-to-right:focus,
        .hvr-sweep-to-right:active {
            color: white;
        }
        .hvr-sweep-to-right:hover:before,
        .hvr-sweep-to-right:focus:before,
        .hvr-sweep-to-right:active:before {
            -webkit-transform: scaleX(1);
            transform: scaleX(1);
        }
    </style>

</head>

<body>
    <a href="#" class="hvr-sweep-to-right">sweep</a>
</body>

</html>
```

### PR DESCRIPTION
**Pull request by mrobinson**
This change extends the DocumentAnimationSet to hold animations for
pseudo-elements. Since pseudo-elements in Servo are not in the DOM like
in Gecko, they need to be handled a bit carefully in stylo.  When a
pseudo-element has an animation, recascade the style. Finally, this
change passes the pseudo-element string properly to animation events.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #10316
- [x] There are tests for these changes

### CODE REVIEW

File: components/style/animation.rs [hunk 1]
```diff
@@ -1138,7 +1139,7 @@ impl ElementAnimationSet {
 
 #[derive(Clone, Debug, Eq, Hash, MallocSizeOf, PartialEq)]
 /// A key that is used to identify nodes in the `DocumentAnimationSet`.
-pub struct AnimationSetKey(pub OpaqueNode);
+pub struct AnimationSetKey(pub OpaqueNode, pub Option<PseudoElement>);
```
Comment (jdm) [inline components/style/animation.rs:1142]: Perhaps this should be a struct with named members instead now.

File: components/layout_thread_2020/dom_wrapper.rs [hunk 2]
```diff
@@ -484,7 +484,7 @@ impl<'le> TElement for ServoLayoutElement<'le> {
         let node = self.as_node();
         let document = node.owner_doc();
         context.animations.get_animation_declarations(
-            &AnimationSetKey(node.opaque()),
+            &AnimationSetKey(node.opaque(), None /* pseudo_element */),
```
Comment (jdm) [inline components/layout_thread_2020/dom_wrapper.rs:487]: Perhaps we should have a `AnimationSetKey::non_pseudo` constructor instead.

File: tests/wpt/metadata/css/css-animations/animationevent-pseudoelement.html.ini [hunk 3]
```diff
@@ -1,6 +1,2 @@
 [animationevent-pseudoelement.html]
   bug: https://github.com/servo/servo/issues/10316
```
Comment (jdm) [inline tests/wpt/metadata/css/css-animations/animationevent-pseudoelement.html.ini:2]: This file (and similar ones referencing 10316) can be removed now.

File: tests/wpt/metadata/css/css-transitions/pseudo-elements-001.html.ini [hunk 4]
```diff
@@ -2,12 +2,6 @@
   [transition padding-left on :before / values]
```
Comment (jdm) [inline tests/wpt/metadata/css/css-transitions/pseudo-elements-001.html.ini:2]: What does it mean that this subtest fails?

File: components/style/matching.rs [hunk 5]
```diff
@@ -464,57 +474,149 @@ trait PrivateMatchMethods: TElement {
...
+        use crate::animation::AnimationSetKey;
+        use crate::dom::TDocument;
+
+        let style_changed = self.process_animations_for_style(
             context,
-            old_values,
+            &mut old_styles.primary,
             new_resolved_styles.primary_style_mut(),
-        ) {
+            None, /* psuedo_element */
```
Comment (jdm) [inline components/style/matching.rs:489]: nit: pseudo_element

File: components/layout_2020/flow/root.rs [hunk 6]
```diff
@@ -446,11 +448,15 @@ impl FragmentTree {
         })
     }
 
-    pub fn remove_nodes_in_fragment_tree_from_set(&self, set: &mut FxHashSet<OpaqueNode>) {
+    pub fn remove_nodes_in_fragment_tree_from_set(&self, set: &mut FxHashSet<AnimationSetKey>) {
         self.find(|fragment, _| {
-            if let Some(tag) = fragment.tag().as_ref() {
-                set.remove(&tag.node());
-            }
+            let (node, pseudo) = match fragment.tag() {
```
Comment (emilio) [inline components/layout_2020/flow/root.rs:453]: nit: can you use `match fragment.tag()? { Tag::Node(...) => { .. } }`?

File: components/style/animation.rs [hunk 7]
```diff
@@ -1143,11 +1137,137 @@ impl ElementAnimationSet {
...
+    pub sets: Arc<RwLock<FxHashMap<AnimationSetKey, ElementAnimationSet>>>,
+}
+
+impl DocumentAnimationSet {
+    /// Return whether or not the provided node has active CSS animations.
+    pub fn has_active_animations(&self, key: &AnimationSetKey) -> bool {
+        self.sets
+            .read()
+            .get(key)
+            .map(|set| set.has_active_animation())
```
Comment (emilio) [inline components/style/animation.rs:1158]: map_or, here and elsewhere?

File: components/style/matching.rs [hunk 8]
```diff
@@ -401,14 +405,20 @@ trait PrivateMatchMethods: TElement {
         // in addition to the unvisited styles.
 
         let mut tasks = UpdateAnimationsTasks::empty();
-        if self.needs_animations_update(context, old_values.as_ref().map(|s| &**s), new_values) {
+        if self.needs_animations_update(
+            context,
+            old_values.as_ref().map(|s| &**s),
+            new_values,
+            None, /* pseudo_element */
```
Comment (emilio) [inline components/style/matching.rs:412]: I think the most common syntax is `/* pseudo_element = */ None,` but again no big deal.

File: components/style/style_resolver.rs [hunk 9]
```diff
@@ -298,6 +298,28 @@ where
...
+            };
+
+            self.cascade_style_and_visited(
+                inputs,
+                Some(primary_style.style()),
+                layout_parent_style_for_pseudo,
+                Some(pseudo),
+            )
+        })
+    }
```
Comment (emilio) [inline components/style/style_resolver.rs:322]: nit: newline missing.

File: components/style/style_resolver.rs [hunk 10]
```diff
@@ -298,6 +298,28 @@ where
...
 
+    /// Cascade a set of rules for pseudo element, using the default parent for inheritance.
+    pub(crate) fn cascade_style_and_visited_for_pseudo_with_default_parents(
+        &mut self,
+        inputs: CascadeInputs,
+        pseudo: &PseudoElement,
+        primary_style: &PrimaryStyle,
+    ) -> ResolvedStyle {
+        with_default_parent_styles(self.element, |_, layout_parent_style| {
+            let layout_parent_style_for_pseudo = if primary_style.style().is_display_contents() {
```
Comment (emilio) [inline components/style/style_resolver.rs:309]: Can we reuse the logic here somehow? I guess it's not too much of an issue, but I think I'd prefer a `fn layout_parent_style_for_pseudo(&PrimaryStyle, Option<&ComputedStyle>)` or what not

File: components/style/animation.rs (hunk 1 above)
Comment (mrobinson) [reply components/style/animation.rs:1142]: Sure. This definitely improves some of the code in `script/animations.rs`.

File: components/layout_thread_2020/dom_wrapper.rs (hunk 2 above)
Comment (mrobinson) [reply components/layout_thread_2020/dom_wrapper.rs:487]: This is a great idea. I've added three constructors `new`, `new_for_non_pseudo`, and `new_for_pseudo`.

File: tests/wpt/metadata/css/css-animations/animationevent-pseudoelement.html.ini (hunk 3 above)
Comment (mrobinson) [reply tests/wpt/metadata/css/css-animations/animationevent-pseudoelement.html.ini:2]: Thanks. I've removed these.

File: tests/wpt/metadata/css/css-transitions/pseudo-elements-001.html.ini (hunk 4 above)
Comment (mrobinson) [reply tests/wpt/metadata/css/css-transitions/pseudo-elements-001.html.ini:2]: It looks like this test trigger a bug in Layout 2013, where pseudo content that has an empty content string is not represented properly in the flow tree. I've confirmed that the tests pass when these test cases use a non-empty content string.

File: components/style/matching.rs (hunk 5 above)
Comment (mrobinson) [reply components/style/matching.rs:489]: :+1:

File: components/layout_2020/flow/root.rs (hunk 6 above)
Comment (mrobinson) [reply components/layout_2020/flow/root.rs:453]: Sure. This is a lot nicer.

File: components/style/animation.rs (hunk 7 above)
Comment (mrobinson) [reply components/style/animation.rs:1158]: I've fixed this here and also one instance in the script crate as well.

File: components/style/matching.rs (hunk 8 above)
Comment (mrobinson) [reply components/style/matching.rs:412]: :+1:

File: components/style/style_resolver.rs (hunk 9 above)
Comment (mrobinson) [reply components/style/style_resolver.rs:322]: Fixed. It's curious that tidy didn't correct this.

File: components/style/style_resolver.rs (hunk 10 above)
Comment (mrobinson) [reply components/style/style_resolver.rs:309]: For sure. I've added a new helper function. Let me know if it's what you were thinking.

**Comment by mrobinson** [comment]
Thanks for the reviews. I've incorporated your suggested change and this PR should be ready to review and merge whenever.

File: components/style/animation.rs [hunk 11]
```diff
@@ -1202,6 +1233,65 @@ impl DocumentAnimationSet {
...
+    pub(crate) fn cancel_all_animations_for_key(&self, key: &AnimationSetKey) {
+        if let Some(set) = self.sets.write().get_mut(key) {
+            set.cancel_all_animations();
+        }
+    }
+}
+
+/// A set of property declarations for a node, including animations and
+/// transitions.
+pub struct AnimationAndTransitionDeclarations {
```
Comment (emilio) [inline components/style/animation.rs:1275]: This looks a lot like [this pre-existing structure](https://searchfox.org/mozilla-central/rev/027893497316897b8f292bde48dbb6da2391a331/servo/components/style/properties/declaration_block.rs#31-38).

It doesn't seem like its usage it's too wide, and I prefer this, so let's unify them in a follow-up? :)

File: components/style/animation.rs [hunk 12]
```diff
@@ -1202,6 +1233,65 @@ impl DocumentAnimationSet {
...
+    /// Declarations for animations.
+    pub animations: Option<Arc<Locked<PropertyDeclarationBlock>>>,
+    /// Declarations for transitions.
+    pub transitions: Option<Arc<Locked<PropertyDeclarationBlock>>>,
+}
+
+impl AnimationAndTransitionDeclarations {
+    /// Create an empty `AnimationAndTransitionDeclarations`.
```
Comment (emilio) [inline components/style/animation.rs:1283]: You could use `#[derive(Default)]` for this if you wanted, fwiw.

File: components/style/style_resolver.rs [hunk 13]
```diff
@@ -118,6 +118,17 @@ where
     )
 }
 
+fn layout_parent_style_for_pseudo<'a>(
```
Comment (emilio) [inline components/style/style_resolver.rs:121]: Yeah, this looks great, thanks!

File: components/style/servo/selector_parser.rs [hunk 14]
```diff
@@ -29,7 +29,7 @@ use style_traits::{ParseError, StyleParseErrorKind};
 /// A pseudo-element, both public and private.
 ///
 /// NB: If you add to this list, be sure to update `each_simple_pseudo_element` too.
-#[derive(Clone, Debug, Eq, Hash, MallocSizeOf, PartialEq, ToShmem)]
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, MallocSizeOf, PartialEq, Serialize, ToShmem)]
```
Comment (emilio) [inline components/style/servo/selector_parser.rs:32]: I don't _think_ you're using the `Copy` here, but ICBW.

**Review by emilio — APPROVED** [review]
This looks pretty sensible to me, with a follow-up PR (or at least issue on file) to unify those structs. Thanks!

File: components/style/animation.rs (hunk 11 above)
Comment (mrobinson) [reply components/style/animation.rs:1276]: Sure. Not a problem at all. I'll make a PR for this once this one lands.

File: components/style/animation.rs (hunk 12 above)
Comment (mrobinson) [reply components/style/animation.rs]: :+1:

File: components/style/servo/selector_parser.rs (hunk 14 above)
Comment (mrobinson) [reply components/style/servo/selector_parser.rs:32]: I just double-checked. For some reason this is necessary in a few places.

**Comment by mrobinson** [comment]
@bors-servo r=emilio

Thanks again for the reviews!

**Comment by mrobinson** [comment]
@bors-servo retry
  *  #26895

**Pull request merged**